File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        #finding middle
        slow=head
        fast=head.next
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next

        #reverse the second half
        start = slow.next
        prev = slow.next = None
        while start:
            tmp = start.next
            start.next = prev
            prev = start
            start = tmp
        
        #merging both
        first, second =head, prev
        while second:
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first = tmp1
            second = tmp2
            

File: Data Structures & Algorithms/reverse-string/submission-1.java
class Solution {
    public void reverseString(char[] s) {
        int n = s.length;
        int i = 0;
        int j = n-1;
        while(i<j){
            char first = s[i];
            s[i] = s[j];
            s[j] = first;
            j--;
            i++;
        }
    }
}